dmq-node/src/DMQ/NodeToClient/LocalMsgSubmission.hs

```diff
@@ -1,64 +1,94 @@
-{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE FlexibleContexts     #-}
+{-# LANGUAGE OverloadedStrings    #-}
+{-# LANGUAGE StandaloneDeriving   #-}
+{-# LANGUAGE TypeApplications     #-}
+{-# LANGUAGE UndecidableInstances #-}
 
 module DMQ.NodeToClient.LocalMsgSubmission where
 
 import Control.Concurrent.Class.MonadSTM
+import Control.Monad.Class.MonadThrow
 import Control.Tracer
 import Data.Aeson (ToJSON (..), object, (.=))
 import Data.Aeson qualified as Aeson
-import Data.Maybe
-
-import Ouroboros.Network.TxSubmission.Inbound.V2
+import Data.Typeable
 
 import DMQ.Protocol.LocalMsgSubmission.Server
 import DMQ.Protocol.LocalMsgSubmission.Type
+import Ouroboros.Network.TxSubmission.Mempool.Simple
 
 -- | Local transaction submission server, for adding txs to the 'Mempool'
 --
 localMsgSubmissionServer ::
-     MonadSTM m
+     forall msgid msg idx m.
+     ( MonadSTM m
+     , MonadThrow m
+     , Typeable msgid
+     , Typeable msg
+     , Show msgid
+     , Show (TxValidationFail msg))
   => (msg -> msgid)
   -- ^ get message id
-  -> Tracer m (TraceLocalMsgSubmission msgid)
-  -> TxSubmissionMempoolWriter msgid msg idx m
+  -> Tracer m (TraceLocalMsgSubmission msg msgid)
+  -> MempoolWriter msgid msg idx m
+  -- ^ duplicate error tag in case the mempool returns the empty list on failure
   -> m (LocalMsgSubmissionServer msg m ())
-localMsgSubmissionServer getMsgId tracer TxSubmissionMempoolWriter { mempoolAddTxs } =
+localMsgSubmissionServer getMsgId tracer MempoolWriter { mempoolAddTxs } =
     pure server
   where
-    failure =
-      -- TODO remove dummy hardcode when mempool returns reason
-      (SubmitFail SigExpired, server) <$ traceWith tracer (TraceSubmitFailure SigExpired)
-    success msgid =
+    process (Left (msgid, reason)) = do
+      traceWith tracer (TraceSubmitFailure msgid reason)
+      throwIO $ MsgValidationException msgid reason
```

Comment:
This will throw an exception and shutdown the connection.  The other end, will not receive reason why these sigs are invalid.  Instead of throwing we should just return validation errors to the client, reject them from adding them to the mempool.  The connection can continue be operational.


However, on the `node-to-node` protocol, we need throw and close the connection if the signature is invalid.